Add documentation for invoker commands on dialog elements (#42822)
mdn/content: files/en-us/web/html/reference/elements/dialog/index.md

    > [!NOTE]
    > While you can toggle between the open and closed states of non-modal dialog boxes by toggling the presence of the `open` attribute, this approach is not recommended. See {{domxref("HTMLDialogElement.open", "open")}} for more information.

## Description

The HTML `<dialog>` element is used to create both modal and non-modal dialog boxes.
Modal dialog boxes block interaction with the rest of the page, making UI elements "behind" the dialog" inert, while non-modal dialog boxes allow interaction with the rest of the page.

### Controlling dialogs using JavaScript

> JavaScript can be used to show and close the `<dialog>` element.

Reviewer: ```suggestion
JavaScript can be used to display and close the `<dialog>` element.
```

Maybe, for consistency with the rest of the page, and the other docs, like the `HTMLDialogElement` stuff?

Reviewer: I'm Ok with that; I don't think it is a major issue.

Modal dialogs can also be closed by pressing the <kbd>Esc</kbd> key.

### Modal dialogs using invoker commands

Modal dialogs can be declaratively opened and closed using the [Invoker Commands API](/en-US/docs/Web/API/Invoker_Commands_API) HTML attributes [`commandfor`](/en-US/docs/Web/HTML/Reference/Elements/button#commandfor) and [`command`](/en-US/docs/Web/HTML/Reference/Elements/button#command) defined on {{htmlelement("button")}} elements.

The `command` attribute sets the particular command that is to be sent when the `<button>` element is clicked, while `commandfor` sets the `id` of the target dialog.
The commands that can be sent for dialogs are [`"show-modal"`](/en-US/docs/Web/HTML/Reference/Elements/button#show-modal), [`"close"`](/en-US/docs/Web/HTML/Reference/Elements/button#close), and [`"request-close"`](/en-US/docs/Web/HTML/Reference/Elements/button#request-close).

> The HTML below shows how the attributes are applied to button elements so that they open and close the modal dialog with the `id` of "my-dialog".

Reviewer: ```suggestion
The HTML below demonstrates how to apply the attributes to a `<button>` element so it can be pressed to open a modal `<dialog>` with an `id` of "my-dialog".
```

It only opens the `<dialog>`. It can't close it as well.

```html
<button command="show-modal" commandfor="my-dialog">Open dialog</button>

<dialog id="my-dialog">
  <p>This dialog was opened using an invoker command.</p>
  <button commandfor="my-dialog" command="close">Close</button>
</dialog>
```

### Non-modal dialogs using popover commands

> Non-modal dialogs can be declaratively opened, closed, and toggled using the [Popover API](/en-US/docs/Web/API/Popover_API) HTML attributes [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget) and [`popovertargetaction`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertargetaction) defined on {{htmlelement("button")}} or {{htmlelement("input")}} elements.

Reviewer: ```suggestion
Non-modal dialogs can be declaratively opened, closed, and toggled using the [Popover API](/en-US/docs/Web/API/Popover_API) HTML attributes [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget) and [`popovertargetaction`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertargetaction), which can be defined on {{htmlelement("button")}} and {{htmlelement("input")}} elements.
```

The `<dialog>` must be turned into a popover by adding the `popover` attribute.

> You can then use `popovertarget` on a button/input to indicate the target popover, and `popovertargetaction` for the action on the popover when the button is clicked.

Reviewer: ```suggestion
You can then use `popovertarget` on a button/input to indicate the target popover, and `popovertargetaction` to specify the action to occur on the popover when the button is clicked.
```

Note that because the dialog is a popover it will be modal, so you can close it by clicking outside the dialog.

> The HTML below shows how the attributes are applied to button elements so that they show and hide the modal dialog with the `id` of "my-dialog".

Reviewer: ```suggestion
The HTML below shows how to apply the attributes to a `<button>` element so it can be pressed to show and hide a modal `<dialog>` with an `id` of "my-dialog".
```

The `<dialog>` must be turned into a popover by adding the `popover` attribute.
You can then use `popovertarget` on a button/input to indicate the target popover, and `popovertargetaction` to specify the action to occur on the popover when the button is clicked.
Note that, because the dialog is a popover, it will be non-modal, so you can close it by clicking outside the dialog.

The HTML below shows how to apply the attributes to a `<button>` element so it can be pressed to show and hide a modal `<dialog>` with an `id` of "my-dialog".

```html
<button popovertarget="my-dialog">Open dialog</button>

<dialog id="my-dialog" popover>
  <p>This dialog was opened using a popovertargetaction attribute.</p>
  <button popovertarget="my-dialog" popovertargetaction="hide">Close</button>
</dialog>
```

The Popover API also provides properties that can be used to get and set the state in JavaScript.

### Closing dialogs

It is important to provide a closing mechanism for every `dialog` element, and to ensure that this works on devices that might not have a physical keyboard.

There are numerous ways to close a dialog:

> - By submitting the form within the dialog form using the `dialog` method (see the [Using the dialog open attribute](#using_the_dialog_open_attribute) example).

Reviewer: ```suggestion
- Submitting the form within the `<dialog>` element with `method="dialog"` set on the `<form>` element (see the [Using the dialog open attribute](#using_the_dialog_open_attribute) example).
```

just so the reader doesn't get confused and think we are talking about a JS method.

## Examples

### Invoker Command API HTML attributes

This example demonstrates how you can use open and close a modal dialog using the [`commandfor`](/en-US/docs/Web/HTML/Reference/Elements/button#commandfor) and [`command`](/en-US/docs/Web/HTML/Reference/Elements/button#command) HTML attributes of the [Invoker Commands API](/en-US/docs/Web/API/Invoker_Commands_API).

> First we declare a {{htmlelement("button")}} element, setting the `command` to [`"show-modal"`](/en-US/docs/Web/HTML/Reference/Elements/button#show-modal), and the `commandfor` as the `id` of the dialog to open (`my-dialog`).

Reviewer: ```suggestion
First, we declare a {{htmlelement("button")}} element, setting the `command` attribute to [`"show-modal"`](/en-US/docs/Web/HTML/Reference/Elements/button#show-modal), and the `commandfor` attribute to the `id` of the dialog to open (`my-dialog`).
```

This example demonstrates how you can open and close a non-modal dialog using the [`popover`](/en-US/docs/Web/HTML/Reference/Global_attributes/popover), [`popovertarget`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertarget), and [`popovertargetaction`](/en-US/docs/Web/HTML/Reference/Elements/button#popovertargetaction) HTML attributes of the [Popover API](/en-US/docs/Web/API/Popover_API).

Note that the `<dialog>` must be turned into a popover by adding the `popover` attribute, and that you don't need to specify `popovertargetaction` for the button to open the dialog.

```html
<button popovertarget="my-dialog">Open dialog</button>

<dialog id="my-dialog" popover>
  <p>This dialog was opened using a popovertargetaction attribute.</p>
  <button popovertarget="my-dialog" popovertargetaction="hide">Close</button>
</dialog>
```

#### Result

Open the dialog by pressing the "Open dialog" button.
You can close the dialog by selecting the "Close" button or pressing the <kbd>Esc</kbd> key.
You can also close it by selecting outside the dialog, as it is non-modal.

{{EmbedLiveSample("Popover API HTML attributes", "100%", 200)}}

### HTML-only non-modal dialog

> This example demonstrates the creation of a non-modal dialog using only HTML.

Reviewer 2: Fixed by https://github.com/mdn/content/pull/42822/changes#r2801445878

Because of the boolean `open` attribute in the `<dialog>` element, the dialog appears open when the page loads.
The dialog can be closed by clicking the "OK" button because the `method` attribute in the `<form>` element is set to `"dialog"`.

> In this case, no JavaScript is needed to close the form.

Reviewer 2: ```suggestion
### Using the dialog `open` attribute

This example demonstrates how you can set the boolean `open` attribute on a `<dialog>` element, in order to create an HTML-only non-modal dialog that is already open when the page loads.

The dialog can be closed by clicking the "OK" button because the `method` attribute in the `<form>` element is set to `"dialog"`.
In this case, no JavaScript is needed to close the form.
```

> The above examples demonstrate the following three methods of closing modal dialogs:

Reviewer: This text is a bit confusing. For a start, doesn't this example on its own demonstrate all these three possible methods? If not, which examples exactly demonstrate these three methods?

Reviewer 2: It is indeed. I didn't update this section and it refers to previous examples. Essentially all it was trying to do was shown that there are lots of ways that you can close a dialog and that it is important to provide such a method.

To address this I add a new section on closing dialogs up in the guide. From this stuff, I just took the comments about the specific example and put them in the description of the code above.

See https://github.com/mdn/content/pull/42822/changes/4c633488ca4fef7e9892a904e0cfb232bb4c976b
